crates/autopilot/src/database/onchain_order_events.rs

```diff
@@ -276,6 +282,8 @@ impl<T: Send + Sync + Clone, W: Send + Sync> OnchainOrderParser<T, W> {
         Vec<(database::events::EventIndex, OnchainOrderPlacement)>,
         Vec<Order>,
     )> {
+        let block_number_timestamp_hashmap =
+            get_block_numbers_of_events(&self.web3, &events).await?;
```

Comment:
Another way to implement this would be to create the hashmap here and fill in the block numbers you need on demand. So in the loop below you would check if the hashmap already has the block number in which case you do nothing, or if it doesn't you would retrieve it and store it.

Downside of that approach is no parallelism. I think you need block numbers for all events anyway so the current solution is better.